File: Leetcode/string/1178_NumberOfValidWordsForEachPuzzles/Solution.py
class Solution(object):
    def findNumOfValidWords(self, words, puzzles):
        """
        :type words: List[str]
        :type puzzles: List[str]
        :rtype: List[int]
        """
        """
        #TLE solution
        res = []
        for puzzle in puzzles:
            cnt = 0
            for word in words:
                if word.count(puzzle[0])>0 and all(map(puzzle.count, word)):
                    cnt+=1
            res.append(cnt)
        return res
        """
        count = collections.Counter(frozenset(w) for w in words)
        res = []
        for p in puzzles:
            cur = 0
            for k in xrange(len(p)):
                for c in itertools.combinations(p[1:], k):
                    cur += count[frozenset(tuple(p[0]) + c)]
            res.append(cur)
        return res

File: Leetcode/string/1178_NumberOfValidWordsForEachPuzzles/Solution.java
class Solution {
    public List<Integer> findNumOfValidWords(String[] words, String[] puzzles) {
        int[][] puz = new int[puzzles.length][2];
        List<Integer> res = new ArrayList<>();
        int[] word = new int[words.length];
        for(int i = 0; i<puzzles.length; i++){
            puz[i][0] = (int)(puzzles[i].charAt(0)-'a');
            for(int j = 0; j<puzzles[i].length(); j++){
                int bit = 1<<(puzzles[i].charAt(j)-'a');
                puz[i][1]|= bit;
            }
            System.out.println(puz[i][0]+" "+puz[i][1]);
        }
        for(int i = 0; i<words.length; i++){
            for(int j = 0; j<words[i].length(); j++){
                int bit = 1 << (words[i].charAt(j)-'a');
                word[i]|=bit;
            }
            System.out.println(word[i]);
        }
        for(int i = 0; i<puz.length; i++){
            int count = 0;
            for(int j = 0; j<word.length; j++){
                if( ( puz[i][1]|word[j] )!= puz[i][1]  ) continue;
                if( (word[j]&(1<<puz[i][0]))!=0 )
                    count+=1;
            }
            res.add(count);
        }
        return res;
    }
}
